mesa-quilt

```diff
@@ -18,17 +18,73 @@
 # 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 
 d=. ; while [ ! -d "$d/nx-X11/extras" ] && [ "$(readlink -e "$d")" != "/" ]; do d="$d/.."; done
-if [ -d "$d/nx-X11/extras" ] && [ -z "$QUILT_PATCHES" ]; then
-    # if in nx-libs tree with unset $QUILT_PATCHES
-    export QUILT_PATCHES="../Mesa.patches"
-    export QUILT_PATCH_OPTS="--reject-format=unified"
-    export QUILT_DIFF_ARGS="-p ab --no-timestamps --no-index --color=auto"
-    export QUILT_REFRESH_ARGS="-p ab --no-timestamps --no-index"
-    export QUILT_COLORS="diff_hdr=1;32:diff_add=1;34:diff_rem=1;31:diff_hunk=1;33:diff_ctx=35:diff_cctx=33"
-    if ! [ -d "$d/nx-X11/extras/Mesa.patches" ]; then mkdir "$d/nx-X11/extras/Mesa.patches"; fi
-    cd "$d/nx-X11/extras/Mesa/"
-fi
 
-quilt "$@"
+if [ -d "$d/nx-X11/extras" ]; then
+
+	if which quilt 1>/dev/null; then
+
+		echo "$(basename $0): quilt utility found, so we use it..."
+		echo
+
+		# if in nx-libs tree with unset $QUILT_PATCHES
+		export QUILT_PATCHES="../Mesa.patches"
+		export QUILT_PATCH_OPTS="--reject-format=unified"
+		export QUILT_DIFF_ARGS="-p ab --no-timestamps --no-index --color=auto"
+		export QUILT_REFRESH_ARGS="-p ab --no-timestamps --no-index"
+		export QUILT_COLORS="diff_hdr=1;32:diff_add=1;34:diff_rem=1;31:diff_hunk=1;33:diff_ctx=35:diff_cctx=33"
+		if ! [ -d "$d/nx-X11/extras/Mesa.patches" ]; then mkdir "$d/nx-X11/extras/Mesa.patches"; fi
+
+		cd "$d/nx-X11/extras/Mesa/"
+
+		quilt "$@"
+
+		cd - 1> /dev/null
+
+
+	else
+
+		echo "$(basename $0): no quilt utility found, falling back to stupid patch appliance..."
+		echo
+
+		CMDLINE_ARGS="$@"
+
+		# we are really stupid here, we only under stand 'push -a' and 'pop -a'
+		# this should be sufficient for the build process...
 
-cd - 1> /dev/null
+		cd "$d/nx-X11/extras/Mesa/"
+
+		touch "../Mesa.patches/.applied-patches"
+
+		if [ "$CMDLINE_ARGS" = "push -a" ]; then
+
+			cat "../Mesa.patches/series" | grep -E -v "^#.*" | while read patchfile; do
+				if [ -n "$patchfile" ] && [ -r "../Mesa.patches/$patchfile" ] && ! cat ../Mesa.patches/.applied-patches | grep -q $patchfile; then
+					echo "Applying patch ../Mesa.patches/$patchfile"
+					patch -p1 < "../Mesa.patches/$patchfile"
+					echo $patchfile >> ../Mesa.patches/.applied-patches
+					echo
+				fi
+			done
+
+		elif [ "$CMDLINE_ARGS" = "pop -a" ]; then
+
+			tac "../Mesa.patches/.applied-patches" | while read patchfile; do
+				if [ -n "$patchfile" ] && [ -r "../Mesa.patches/$patchfile" ]; then
+					echo "Removing patch ../Mesa.patches/$patchfile"
+					patch -R -p1 < "../Mesa.patches/$patchfile"
+					sed -i "/$patchfile/d" "../Mesa.patches/.applied-patches"
```

Comment:
> One more thing: I prefer compact shell scripts, this one has far too many empty lines for me. But YMMV

@uli42: /me prefers readable scripts to compact scripts. I consider bash more as a progamming language than an admin tool, actually.